PeDuCoML/lib/parser.ml

```diff
@@ -100,28 +58,131 @@ let parse_capitalized_entity =
   else fail "Parsing error: not a capitalized entity."
 ;;
 
-let data_constructors = [ "Ok"; "Error"; "Some"; "None" ]
-
 let keywords =
-  [ "let"
-  ; "rec"
-  ; "match"
-  ; "with"
-  ; "if"
-  ; "then"
-  ; "else"
-  ; "in"
-  ; "fun"
-  ; "and"
-  ; "effect"
-  ; "type"
-  ; "perform"
-  ; "continue"
-  ]
+  [ "let"; "rec"; "match"; "with"; "if"; "then"; "else"; "in"; "fun"; "and"; "type" ]
 ;;
 
 (* ------- *)
 
+(* Patterns parsers *)
+
+let parse_pattern d =
+  choice
+    [ d.parse_ptuple d
+    ; d.parse_pconstruct_list d
+    ; d.parse_plist d
+    ; d.parse_pliteral
+    ; d.parse_pidentifier
+    ; d.parse_wild
+    ]
+;;
+
+let parse_pliteral =
+  fix
+  @@ fun self ->
+  remove_spaces
+  *> (parens self
+      <|>
+      let is_digit = function
+        | '0' .. '9' -> true
+        | _ -> false
+      in
+      let parse_int_literal =
+        take_while1 is_digit >>| int_of_string >>| fun x -> LInt x
+      in
+      let parse_string_literal =
+        char '"' *> take_while (( != ) '"') <* char '"' >>| fun x -> LString x
+      in
+      let parse_char_literal = char '\'' *> any_char <* char '\'' >>| fun x -> LChar x in
+      let parse_bool_literal =
+        string "true" <|> string "false" >>| bool_of_string >>| fun x -> LBool x
+      in
+      let parse_unit_literal = string "()" >>| fun _ -> LUnit in
+      let parse_literal =
+        choice
+          [ parse_int_literal
+          ; parse_string_literal
+          ; parse_char_literal
+          ; parse_bool_literal
+          ; parse_unit_literal
+          ]
+      in
+      pliteral <$> parse_literal)
```

Comment:
Hmm, I look at this function and `parse_pliteral`. The only difference is the last line. I would like us to avoid code duplication wherever possible. Perhaps it would be better to do smth like this:
`let parse_a_literal smart_constructor =`
`    ...`
`    in`
`    smart_constructor <$> parse_literal)`
And then you can write as follows:
`let parse_pliteral = parse_a_literal pliteral`
`let parse_eliteral = parse_a_literal eliteral`